plugins/events/eventrouter.go

```diff
@@ -0,0 +1,209 @@
+)
+
+var Log = log.WithField("system", "events")
+var leadershipName = "wf_collector_events"
+var filteredEvents = gometrics.GetOrRegisterCounter("events.filtered", gometrics.DefaultRegistry)
+var receviedEvents = gometrics.GetOrRegisterCounter("events.recevied", gometrics.DefaultRegistry)
+var sentEvents = gometrics.GetOrRegisterCounter("events.sent", gometrics.DefaultRegistry)
+
+type EventRouter struct {
+	kubeClient        kubernetes.Interface
+	eLister           corelisters.EventLister
+	eListerSynched    cache.InformerSynced
+	sink              metrics.DataSink
+	sharedInformers   informers.SharedInformerFactory
+	stop              chan struct{}
+	daemon            bool
+	leadershipManager *leadershipManager
+	whitelist         map[string]glob.Glob
+	blacklist         map[string]glob.Glob
+}
+
+func CreateEventRouter(clientset kubernetes.Interface, cfg *configuration.EventsConfig, sink metrics.DataSink, clusterName string, daemon bool) *EventRouter {
+	sharedInformers := informers.NewSharedInformerFactory(clientset, time.Minute)
+	eventsInformer := sharedInformers.Core().V1().Events()
+
+	er := &EventRouter{
+		kubeClient:      clientset,
+		sink:            sink,
+		daemon:          daemon,
+		sharedInformers: sharedInformers,
+		whitelist:       filter.MultiCompile(cfg.Filters.TagWhitelist),
+		blacklist:       filter.MultiCompile(cfg.Filters.TagBlacklist),
+	}
+
+	eventsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: er.addEvent,
+	})
+	er.eLister = eventsInformer.Lister()
+	er.eListerSynched = eventsInformer.Informer().HasSynced
+
+	if er.daemon {
+		er.leadershipManager = newLeadershipManager(er, leadershipName, clientset)
+	}
+	return er
+}
+
+func (er *EventRouter) Start() {
+	if er.daemon {
+		er.leadershipManager.Start()
+	} else {
+		go func() { er.resume() }()
+	}
+}
+
+func (er *EventRouter) resume() {
+	er.stop = make(chan struct{})
+	defer utilruntime.HandleCrash()
+
+	Log.Infof("Starting EventRouter")
+
+	go func() { er.sharedInformers.Start(er.stop) }()
+
+	// here is where we kick the caches into gear
+	if !cache.WaitForCacheSync(er.stop, er.eListerSynched) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+	<-er.stop
+
+	Log.Infof("Shutting down EventRouter")
+}
+
+func (er *EventRouter) pause() {
+	close(er.stop)
+}
+
+func (er *EventRouter) Stop() {
+	if er.daemon {
+		er.leadershipManager.Stop()
+	}
+	er.pause()
+}
+
+// addEvent is called when an event is created, or during the initial list
+func (er *EventRouter) addEvent(obj interface{}) {
+	e := obj.(*v1.Event)
+
+	ns := e.InvolvedObject.Namespace
+	if len(ns) == 0 {
+		ns = "default"
+	}
+
+	tags := map[string]string{
+		"namespace_name": ns,
+		"kind":           e.InvolvedObject.Kind,
+		"reason":         e.Reason,
+		"component":      e.Source.Component,
+	}
+
+	receviedEvents.Inc(1)
+	if len(er.whitelist) > 0 && !filter.MatchesTags(er.whitelist, tags) {
```

Comment:
Refactor the filtering part into a separate method and add unit tests around them.

Reply:
`filter.MatchesTags` already have tests.